### DISCUSSION TITLE
Text Truncation

### DISCUSSION THREAD

**Discussion by girlpunk** [QuestPDF/QuestPDF#742] [Ideas]
Likely one for when the Skia features are finished, but a handy text option would be to truncate text to fix a space.

For example:
![image](https://github.com/QuestPDF/QuestPDF/assets/161093/b4d8edad-a712-46d5-a2f0-ce04dc51e16b)
This is currently being done manually using Humanizer's truncate function, which is limited to truncating to a fixed number of characters. As such, a guess must be made as to the amount of text that will fit.
```csharp
container
    .AlignRight()
    .AlignBottom()
    .Text(
        title
            .Truncate(40, "…")
            .Transform(To.UpperCase)
    )
    .Style(styles.H5);
```
This results is some potential issues given variable character widths. The option to do something similar directly within QuestPDF would solve these problems.

(note that the lower-case text in the crosshatched area is unrelated)

**Comment by MarcinZiabek** [comment]
I agree!

Right now, it can be solved by using the `MaxHeight()` API (that is high enough to fit one line of text but low enough not to allow line breaking) in combination with the `StopPaging()` API.

The SkParagraph module in Skia should help to support similar cases natively with much higher flexibility :)

**Reply by girlpunk** [reply]
Would using `MaxHeight()` allow including truncation characters, i.e. `no decaf in coffee mac...`?

**Reply by MarcinZiabek** [reply]
Unfortunately not  😥

However, it is a part of SkParagraph, and I would like to port it as a part of the ParagraphStyle class within the custom-skin build effort. https://github.com/QuestPDF/QuestPDF/discussions/622

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Good news — the feature requested here has since shipped. Text can now be clamped to a fixed number of lines with a customizable ellipsis via [ClampLines](https://www.questpdf.com/api-reference/text/paragraph-style.html#clamp-line-with-ellipsis), so clamping to a single line with .ClampLines(1, "…") produces exactly the truncated result you described. As it is width-aware, it also avoids the variable-character-width issues of a fixed character count.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
